### ISSUE TITLE
Contributions accepted?

### ISSUE THREAD

**Issue by masci** [ManderaGeneral/generalimport#18]
Very nice project, congrats!

I see in the [project metadata](https://github.com/ManderaGeneral/generalimport/blob/master/metadata.json#L12) that `contribute` is set to `False` - is this going to change in the future?

**Comment by Mandera** [comment]
Thank you! I'm stoked you find it interesting!

Yes, that's the plan! There are a few things I need to decide on beforehand, all the legal stuff is a bit overwhelming

- I'm considering changing the license to Apache for the patent clause to make adoption easier for companies
- I need to make sure there aren't any dependencies with a copyleft license
- I'm pretty sure I don't want a CLA - It seems to be too intrusive
  - That means any contributor will keep their copyright, which means changing the license will require the approval of all copyright holders
- Then there's the issue of making sure contributions are made by the actual contributor, which means I probably do want to require a DCO

**Comment by Mandera** [comment]
Thank you for your interest and contributions @ZanSara! Please see this issue - I've moved it up my priority list so please stay tuned.

I will get to your contributions once this is complete. I'd like to apologize for the inconvenience, you guys are the first contributors for me!

### Current state
- I will change the license to Apache 2
- I'm still torn on whether to require signed commits for a [DCO](https://developercertificate.org/)
  - The goal is simply to ensure that contributors comply with the Apache 2 license
  - [Section 7 of this article](https://opensource.guide/legal/) makes it sound like it's not necessary, which would be great!

**Comment by ZanSara** [comment]
Hey @Mandera! Happy to help :blush: I really like the approach you're taking with this library so I'm more than willing to contribute bugfixes and small features.

Me and @masci would love to use this library to manage optional imports in [Haystack](https://github.com/deepset-ai/haystack), which has tons of dependencies and has always struggled with how to manage this growing list effectively. So you can expect some activity from us (and maybe other colleagues of ours) in the next weeks. Apart from the management of exceptions, I know already that some of our most complex dependencies (specifically `sqlalchemy`, but there are surely others) currently don't work with `generalimport`, so I will contribute with the bug's descriptions as issues, and later bugfixes for those as soon as I have them. I hope you don't mind me being a bit proactive and going ahead with bugfixes as soon as I need them.

Depending on how you feel about this, I am also willing to work some more "contributor friendly" features: linting and code formatting, pre-commit hooks, API docs on GitHub pages,  a `pyproject.toml` file, a CI with tests, and so on. I will open a separate issues where I can propose some ideas on this front and you can choose which ones you like and which ones you don't.

Regarding the license, for now either (MIT, Apache2) is fine. Regarding copyrights, for us a CLA would work if that simplifies your life (in Haystack we have a bot that asks contributors to sign it and it works quite well), but I'm happy to wait for your decision either way :slightly_smiling_face:

**Comment by Mandera** [comment]
Wow that's awesome! I'm thrilled you're looking to implement generalimport!

Most of these things would go through generalpackager which is my custom build system that runs the tests and synchronizes everything upon success. You can see the entry points [here ](https://github.com/ManderaGeneral/generalimport/blob/ea3e12247fe0512a280b01808966d1996820d94f/.github/workflows/workflow.yml#L48)in the workflow `workflow_unittest` and `workflow_sync`.

- All the files it handles are [here](https://github.com/ManderaGeneral/generalpackager/tree/master/generalpackager/api/shared/files/definitions) - Some of them are only for defining the path
- Linting would be great! It's not something I've dabbled with yet, is it related to [black](https://github.com/ManderaGeneral/generalpackager/issues/47)? If not, which project do you propose to use?
- Replacing `setup.py` with `pyproject.toml` would be great - started an [issue](https://github.com/ManderaGeneral/generalpackager/issues/78)
- I've got a pre-push hook defined [here](https://github.com/ManderaGeneral/generalpackager/blob/master/generalpackager/api/shared/files/definitions/pre_push_hook.py), it's generating non-aesthetic files to return non-zero if any of those files are changed (Meant to catch an outdated workflow or setup.py for example)
- API docs on GitHub pages would be amazing, I have no idea how that works though
- The CLA is really interesting!
  - Do you require every single commit to be signed in Haystack? Or is the CLA enough?
  - The [CLA assistant](https://cla-assistant.io/deepset-ai/haystack) seems pretty nice, so is that a one-time sign for every significant contributor?
  - Does the CLA in your example transfer the copyright to Haystack, or does the contributor still keep the copyright?
- I'd love to get to the bottom of why `sqlalchemy` doesn't work!

The current state of generalpackager is not great. I've only been using the master branch so now I'm in the middle of a breaking update in it. I will see if I can move that new stuff to a new branch and create a new workflow like you were talking about too.

Because in the workflow it's cloning every repo to sync it all, so if master has breaking changes it will just fail - not ideal haha!

Github also had to update their SSH keys so that broke the workflow as well, needing [some love](https://github.com/ManderaGeneral/generalpackager/issues/77)!

I'd like to probably protect the master branch and only allow reviewed pull requests, once all tests pass.

Currently every package in ManderaGeneral is tested and synced on every commit (Unless message starts with [CI SKIP]). If any of the versions are bumped then it will bump any package that has had non-aesthetic changes to it too. When a bumped package is being synced it will get [tagged](https://github.com/ManderaGeneral/generalimport/tags) and published to pypi.

I'm planning to put all this information in a CONTRIBUTING.md eventually.

The long-term goal is to generalize generalpackager to work for any environment, not only ManderaGeneral, but that's far away.

To summarize; I'm loving your contribution ideas but a lot of them will have to go through generalpackager

**Comment by ZanSara** [comment]
By the way: I might have found out why `sqlalchemy` caused me trouble: might be not `generalimport`s fault. I'll open an issue only if I'm sure it's due to this library, which now seems more unlikely.

**Comment by Mandera** [comment]
Alright cool! I've made some decisions and issues
- [No CLA](https://ben.balter.com/2018/01/02/why-you-probably-shouldnt-add-a-cla-to-your-open-source-project/) - Since I'm not a company I think it makes sense to reduce friction
- [No DCO](https://strapi.io/blog/switching-from-dco-to-cla) - Too much hassle
- https://github.com/ManderaGeneral/generalpackager/issues/79
- https://github.com/ManderaGeneral/generalpackager/issues/80
- https://github.com/ManderaGeneral/generalpackager/issues/81

**Comment by masci** [comment]
Hey @Mandera thanks a lot for your support!

> Does the CLA in your example transfer the copyright to Haystack, or does the contributor still keep the copyright?

Also for @ZanSara: yes, by signing the CLA contributors transfer their copyright to deepset. This is commonly done to allow companies to easily manage legal disputes or implement a license change without collecting consensus from every single contributor (reason why CLAs can be a show stopper for many potential contributors).

**Comment by Mandera** [comment]
Contributions are enabled and the license has been changed! The other changes can happen periodically
- https://github.com/ManderaGeneral/generalimport/milestone/2
- https://github.com/ManderaGeneral/generalpackager/milestone/7
